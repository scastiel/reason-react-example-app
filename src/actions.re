open Interval;

type action =
  | Increment
  | Decrement
  | StartIncrementing intervalId
  | StopIncrementing;

type deferredAction = (action => unit) => unit;

let incrementAction = Increment;

let decrementAction = Decrement;

let startIncrementingAction intervalId => StartIncrementing intervalId;

let stopIncrementingAction = StopIncrementing;

let increment dispatch => incrementAction |> dispatch;

let decrement dispatch => decrementAction |> dispatch;

let startIncrementing intervalId dispatch => startIncrementingAction intervalId |> dispatch;

let stopIncrementing dispatch => stopIncrementingAction |> dispatch;

let incrementEverySecond dispatch => {
  let intervalId = setInterval (fun () => increment dispatch) 1000;
  startIncrementing intervalId dispatch
};

let stopIncrementingEverySecond intervalId dispatch => {
  clearInterval intervalId;
  stopIncrementing dispatch
};
