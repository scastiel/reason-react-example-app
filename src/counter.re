open Interval;

let component = ReasonReact.statelessComponent "Counter";

let make
    counter::(counter: int)
    intervalId::(intervalId: option intervalId)
    increment::(increment: unit => unit)
    incrementEverySecond::(incrementEverySecond: unit => unit)
    stopIncrementingEverySecond::(stopIncrementingEverySecond: intervalId => unit)
    decrement::(decrement: unit => unit)
    _children => {
  ...component,
  render: fun self =>
    <div>
      (ReasonReact.stringToElement ("Counter: " ^ string_of_int counter))
      <button className="plus-button" onClick=(self.handle (fun _ _ => increment ()))>
        (ReasonReact.stringToElement "+")
      </button>
      (
        switch intervalId {
        | None =>
          <button
            className="plus-1s-button" onClick=(self.handle (fun _ _ => incrementEverySecond ()))>
            (ReasonReact.stringToElement "+ every 1s")
          </button>
        | Some intervalId =>
          <button
            className="stop-plus-1s-button"
            onClick=(self.handle (fun _ _ => stopIncrementingEverySecond intervalId))>
            (ReasonReact.stringToElement "stop +")
          </button>
        }
      )
      <button className="minus-button" onClick=(self.handle (fun _ _ => decrement ()))>
        (ReasonReact.stringToElement "-")
      </button>
    </div>
};

let counter =
  ReasonReact.wrapReasonForJs
    ::component
    (
      fun jsProps =>
        make
          counter::jsProps##counter
          intervalId::jsProps##intervalId
          increment::jsProps##increment
          incrementEverySecond::jsProps##incrementEverySecond
          stopIncrementingEverySecond::jsProps##stopIncrementingEverySecond
          decrement::jsProps##decrement
          [||]
    );
