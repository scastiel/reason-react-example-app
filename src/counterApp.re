open State;

open Actions;

let component = ReasonReact.statelessComponent "CounterApp";

let make state::(state: state) dispatch::(dispatch: deferredAction => unit) _children => {
  ...component,
  render: fun _ => {
    let onIncrement () => dispatch increment;
    let onIncrementEverySecond () => dispatch incrementEverySecond;
    let onStopIncrementingEverySecond intervalId =>
      dispatch (stopIncrementingEverySecond intervalId);
    let onDecrement () => dispatch decrement;
    <Counter
      intervalId=state.intervalId
      counter=state.counter
      increment=onIncrement
      incrementEverySecond=onIncrementEverySecond
      stopIncrementingEverySecond=onStopIncrementingEverySecond
      decrement=onDecrement
    />
  }
};

let counterApp =
  ReasonReact.wrapReasonForJs
    ::component (fun jsProps => make state::jsProps##state dispatch::jsProps##dispatch [||]);
