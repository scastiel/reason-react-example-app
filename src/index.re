open State;

open Reducer;

let createComponent state dispatch => <CounterApp state dispatch />;

module Provider = ProviderFactory.MakeProvider State;

ReactDOMRe.renderToElementWithId
  <Provider reducer initialState=(createState ()) createComponent /> "root";
