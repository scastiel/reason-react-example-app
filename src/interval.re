type intervalId = int;

external setInterval : (unit => unit) => int => intervalId = "window.setInterval" [@@bs.val];

external clearInterval : intervalId => unit = "window.clearInterval" [@@bs.val];
