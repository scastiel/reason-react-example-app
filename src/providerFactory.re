module MakeProvider (State: {type state;}) => {
  type self = ReasonReact.self State.state ReasonReact.noRetainedProps;
  let component = ReasonReact.statefulComponent "Provider";
  let make ::reducer ::initialState ::createComponent _children => {
    let dispatchAction action state => ReasonReact.Update (reducer state action);
    let update (self: self) deferredAction => {
      let onDispatch action => self.update (fun () self => dispatchAction action self.state) ();
      deferredAction onDispatch
    };
    {
      ...component,
      initialState: fun () => initialState,
      render: fun (self: self) => createComponent self.state (update self)
    }
  };
  let provider =
    ReasonReact.wrapReasonForJs
      ::component
      (
        fun jsProps =>
          make
            reducer::jsProps##reducer
            initialState::jsProps##initialState
            createComponent::jsProps##createComponent
            [||]
      );
};
