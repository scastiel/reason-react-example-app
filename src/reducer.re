open State;

open Actions;

type reducer = state => action => state;

let reducer: reducer =
  fun state action =>
    switch action {
    | Increment => {...state, counter: state.counter + 1}
    | Decrement => {...state, counter: state.counter - 1}
    | StartIncrementing intervalId =>
      switch state.intervalId {
      | None => {...state, intervalId: Some intervalId}
      | _ => state
      }
    | StopIncrementing =>
      switch state.intervalId {
      | Some _ => {...state, intervalId: None}
      | _ => state
      }
    };
