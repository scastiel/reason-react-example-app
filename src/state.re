open Interval;

type state = {
  counter: int,
  intervalId: option intervalId
};

let createState () => {counter: 0, intervalId: None};

let setCounter state counter => {...state, counter};

let setIntervalId state intervalId => {...state, intervalId: Some intervalId};

let clearIntervalId state => {...state, intervalId: None};
