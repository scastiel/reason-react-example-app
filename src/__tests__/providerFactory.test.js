import React from 'react'
import { MakeProvider } from '../../lib/js/src/providerFactory'
import CounterApp from '../../lib/js/src/counterApp'
import { shallow } from 'enzyme'
import { createState } from '../../lib/js/src/state'
import { increment, incrementAction } from '../../lib/js/src/actions'

const Provider = MakeProvider()[2]

describe('Provider component', () => {
  it('should render', () => {
    const reducer = jest.fn((state, action) => state)
    const initialState = createState()
    const createComponent = jest.fn((state, dispatch) => <span />)
    const wrapper = shallow(
      <Provider
        reducer={state => action => reducer(state, action)}
        initialState={initialState}
        createComponent={state => dispatch => createComponent(state, dispatch)}
      />
    )
    expect(wrapper).toMatchSnapshot()
    expect(createComponent.mock.calls.length).toEqual(1)
    expect(createComponent.mock.calls[0][0]).toEqual(initialState)
    const dispatch = createComponent.mock.calls[0][1]
    dispatch(d => d(incrementAction))
    expect(reducer.mock.calls.length).toEqual(1)
    expect(reducer.mock.calls[0]).toEqual([initialState, incrementAction])
  })
})
