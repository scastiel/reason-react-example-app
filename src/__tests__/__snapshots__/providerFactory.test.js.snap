// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider component should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <span />,
  "nodes": Array [
    <span />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Provider
        createComponent={[Function]}
        initialState={
                Array [
                        0,
                        0,
                      ]
        }
        reducer={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance":  {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "enqueueMethod": [Function],
        "handleMethod": [Function],
        "props": Object {
          "createComponent": [Function],
          "initialState": Array [
            0,
            0,
          ],
          "reducer": [Function],
        },
        "refs": Object {},
        "self": [Function],
        "state": Object {
          "reasonState": Array [
            0,
            0,
          ],
          "reasonStateVersion": 1,
          "reasonStateVersionUsedToComputeSubelements": 1,
        },
        "transitionNextTotalState": [Function],
        "updateMethod": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <span />,
        "_debugID": 2,
        "_renderedOutput": <span />,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Provider
    createComponent={[Function]}
    initialState={
        Array [
            0,
            0,
          ]
    }
    reducer={[Function]}
/>,
}
`;
